packages/core/utils/src/replaceBundleReferences.js

```diff
@@ -0,0 +1,165 @@
+// @flow strict-local
+
+import type SourceMap from '@parcel/source-map';
+import type {
+  Async,
+  Blob,
+  Bundle,
+  BundleResult,
+  BundleGraph,
+  Dependency
+} from '@parcel/types';
+
+import invariant from 'assert';
+import {Readable} from 'stream';
+import nullthrows from 'nullthrows';
+import URL from 'url';
+import {bufferStream, urlJoin} from '../';
+
+type ReplacementMap = Map<
+  string /* dependency id */,
+  {|from: string, to: string|}
+>;
+
+/*
+ * Replaces references to dependency ids with either:
+ *   - in the case of an inline bundle, the packaged contents of that bundle
+ *   - in the case of another bundle reference, the bundle's url from the publicUrl root
+ *   - in the case of a url dependency that Parcel did not handle,
+ *     the original moduleSpecifier. These are external requests.
+ */
+export async function replaceBundleReferences({
+  bundle,
+  bundleGraph,
+  contents,
+  map,
+  getInlineReplacement,
+  getInlineBundleContents
+}: {|
+  bundle: Bundle,
+  bundleGraph: BundleGraph,
+  contents: string,
+  getInlineReplacement: (
+    Dependency,
+    ?'string',
+    string
+  ) => {|from: string, to: string|},
+  getInlineBundleContents: (
+    Bundle,
+    BundleGraph
+  ) => Async<{|contents: Blob, map: ?(Readable | string)|}>,
+  map?: ?SourceMap
+|}): Promise<BundleResult> {
+  let replacements = new Map();
+
+  for (let {
+    dependency,
+    bundleGroup
+  } of bundleGraph.getBundleGroupsReferencedByBundle(bundle)) {
+    let [entryBundle] = bundleGraph.getBundlesInBundleGroup(bundleGroup);
+    if (entryBundle.isInline) {
+      // inline bundles
+      let packagedBundle = await getInlineBundleContents(
+        entryBundle,
+        bundleGraph
+      );
+      let packagedContents = (packagedBundle.contents instanceof Readable
+        ? await bufferStream(packagedBundle.contents)
+        : packagedBundle.contents
+      ).toString();
+
+      let inlineType = nullthrows(entryBundle.getMainEntry()).meta.inlineType;
```

Comment:
Where do we define what `inlineType` values are allowed? I'm not sure it should be enforced in this util in case new ones need to be defined by plugins.

Reply:
It's not really enforced. `inlineType` isn't a formal concept in core and is just something that transformers and packagers agree on.

Reply:
it's enforced by the line below asserting the values...

Reply:
This function isn't run in core though, it's run in packagers. Perhaps this util shouldn't live in `core/utils`?

Reply:
Yeah true, but the util is still kinda “shared” code that shouldn’t make assumptions if it’s meant to be informal.